<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.34">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Modkit_Analysis_dRNAseq – Gjandebeur Projects</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-c8ad9e5dbd60b7b70b38521ab19b7da4.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-d68be38d83eca2bb035acf846ffba811.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Gjandebeur Projects</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../scripts.html"> 
<span class="menu-text">Scripts</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#drna_analyses" id="toc-drna_analyses" class="nav-link active" data-scroll-target="#drna_analyses">dRNA_analyses</a>
  <ul class="collapse">
  <li><a href="#plot-data-from-filtered-file-coverage-percent-modified" id="toc-plot-data-from-filtered-file-coverage-percent-modified" class="nav-link" data-scroll-target="#plot-data-from-filtered-file-coverage-percent-modified">Plot data from filtered file, coverage &amp; percent modified</a></li>
  <li><a href="#now-running-genomic-ranges-to-split-tx-into-their-genomic-regions" id="toc-now-running-genomic-ranges-to-split-tx-into-their-genomic-regions" class="nav-link" data-scroll-target="#now-running-genomic-ranges-to-split-tx-into-their-genomic-regions">Now running Genomic Ranges to split tx into their genomic regions</a></li>
  <li><a href="#now-running-function-to-find-each-regions-hits" id="toc-now-running-function-to-find-each-regions-hits" class="nav-link" data-scroll-target="#now-running-function-to-find-each-regions-hits">Now running function to find each region’s hits</a></li>
  <li><a href="#now-plotting-genomic-ranges" id="toc-now-plotting-genomic-ranges" class="nav-link" data-scroll-target="#now-plotting-genomic-ranges">Now plotting genomic ranges</a></li>
  <li><a href="#go-enrichment-analyses" id="toc-go-enrichment-analyses" class="nav-link" data-scroll-target="#go-enrichment-analyses">GO Enrichment Analyses</a></li>
  <li><a href="#next-export-the-subset-file-as-bed-to-run-bedtools-intersect-must-be-done-unix-to-line-up-with-reference-file-and-output-as-fasta" id="toc-next-export-the-subset-file-as-bed-to-run-bedtools-intersect-must-be-done-unix-to-line-up-with-reference-file-and-output-as-fasta" class="nav-link" data-scroll-target="#next-export-the-subset-file-as-bed-to-run-bedtools-intersect-must-be-done-unix-to-line-up-with-reference-file-and-output-as-fasta">Next export the subset file as bed to run bedtools intersect (must be done unix) to line up with reference file and output as fasta</a></li>
  <li><a href="#next-code-can-not-be-run-in-rstudio-must-export-to-a-unix-shell" id="toc-next-code-can-not-be-run-in-rstudio-must-export-to-a-unix-shell" class="nav-link" data-scroll-target="#next-code-can-not-be-run-in-rstudio-must-export-to-a-unix-shell">Next code can NOT be run in RStudio, must export to a UNIX shell</a></li>
  <li><a href="#using-pa-lengths-from-original-basecalling-for-analyses" id="toc-using-pa-lengths-from-original-basecalling-for-analyses" class="nav-link" data-scroll-target="#using-pa-lengths-from-original-basecalling-for-analyses">Using pA lengths from original basecalling for analyses</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Modkit_Analysis_dRNAseq</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="drna_analyses" class="level1">
<h1>dRNA_analyses</h1>
<p>Workflow for analyses of m6A Modifications Post Modkit Filtering</p>
<p>load dependencies first</p>
<pre><code>library(data.table)
library(tidyverse) 
library(tidytable)
library(patchwork)
library(ggplot2)
library(GenomicRanges)
library(txdbmaker)</code></pre>
<p>setwd and import m6A_subset from “m6A_subset_allstats.tsv”</p>
<section id="plot-data-from-filtered-file-coverage-percent-modified" class="level3">
<h3 class="anchored" data-anchor-id="plot-data-from-filtered-file-coverage-percent-modified">Plot data from filtered file, coverage &amp; percent modified</h3>
<p>This can and should be scaled up with more modification types and samples</p>
<pre><code>ggplot(m6A_subset, aes(x = avg_n_valid_cov, y = avg_percent_modified, color = mod)) +
  geom_smooth(method = "gam", formula = y ~ s(x, bs = "cs"), se = FALSE, linewidth = 1.2) +
  scale_x_log10(limits = c(10, 10000)) +
  scale_color_brewer(palette = "Set1") +
  labs(
    x = "Valid Coverage (log10)",
    y = "Percent Modified",
    color = "Modification",
    title = "Coverage and Percent Modification"
  ) +
  theme_classic(base_size = 14)</code></pre>
<p><img width="1200" height="800" alt="githubexample_coverage_percentmodified" src="https://github.com/user-attachments/assets/9aa987d0-7218-4c6d-8df4-c3c9030b8554"></p>
<p>load in txdb from gtf file (provided in files)</p>
<pre><code>txdb &lt;- makeTxDbFromGFF("at_ensembl_plants.gtf")
k &lt;- keys(txdb, keytype = "TXNAME")  
tx2gene_at &lt;- AnnotationDbi::select(txdb, keys = k, columns = "GENEID", keytype = "TXNAME")</code></pre>
<p>intersect txdb with the m6A_subset dataframe</p>
<pre><code>#intersecting files
m6A_subset &lt;- m6A_subset %&gt;%
  inner_join(., tx2gene_at, by=c("tx"= "TXNAME")) %&gt;%
  unite(position, chrom:end, remove =F) 

#changing format to make reproducible
m6A_subset &lt;- m6A_subset %&gt;%
  separate(position, into = c("chrom", "start", "end"), sep = "_", remove = FALSE) %&gt;%
  mutate(
    start = as.integer(start),
    end = as.integer(end)
  )</code></pre>
</section>
<section id="now-running-genomic-ranges-to-split-tx-into-their-genomic-regions" class="level3">
<h3 class="anchored" data-anchor-id="now-running-genomic-ranges-to-split-tx-into-their-genomic-regions">Now running Genomic Ranges to split tx into their genomic regions</h3>
<pre><code>mods &lt;- GRanges(
  seqnames = m6A_subset$chrom,
  ranges = IRanges(start = m6A_subset$start, end = m6A_subset$end),
  mod = m6A_subset$mod,
  gene_id = m6A_subset$gene_id
)

cds_gr &lt;- cds(txdb)
five_utr_gr &lt;- fiveUTRsByTranscript(txdb, use.names = TRUE) |&gt; unlist()
three_utr_gr &lt;- threeUTRsByTranscript(txdb, use.names = TRUE) |&gt; unlist()
introns_gr &lt;- intronsByTranscript(txdb, use.names = TRUE) |&gt; unlist()</code></pre>
</section>
<section id="now-running-function-to-find-each-regions-hits" class="level3">
<h3 class="anchored" data-anchor-id="now-running-function-to-find-each-regions-hits">Now running function to find each region’s hits</h3>
<pre><code>assign_region &lt;- function(mods, cds, five_utr, three_utr, introns) {
  region &lt;- rep("intergenic", length(mods))  # default assignment

  hits_cds &lt;- findOverlaps(mods, cds)
  region[queryHits(hits_cds)] &lt;- "CDS"
  hits_5utr &lt;- findOverlaps(mods, five_utr)
  idx_5utr &lt;- setdiff(queryHits(hits_5utr), queryHits(hits_cds))
  region[idx_5utr] &lt;- "5'UTR"  
  hits_3utr &lt;- findOverlaps(mods, three_utr)
  idx_3utr &lt;- setdiff(queryHits(hits_3utr), c(queryHits(hits_cds), idx_5utr))
  region[idx_3utr] &lt;- "3'UTR"
  
  hits_introns &lt;- findOverlaps(mods, introns)
  idx_introns &lt;- setdiff(queryHits(hits_introns), c(queryHits(hits_cds), idx_5utr, idx_3utr))
  region[idx_introns] &lt;- "intron"
  
  return(region)
}

# add region tag to the original dataframe for downstream analyses
m6A_subset$region &lt;- assign_region(mods_gr, cds_gr, five_utr_gr, three_utr_gr, introns_gr)

#Lastly, filter by descending groups 
summary_filtered_data_byregion &lt;-  m6A_subset %&gt;%
  group_by(region, mod) %&gt;%
  summarise(n_mods = n(), .groups = "drop") %&gt;%
  arrange(desc(n_mods))

# For overview of graph
table(m6A_subset$region)</code></pre>
</section>
<section id="now-plotting-genomic-ranges" class="level3">
<h3 class="anchored" data-anchor-id="now-plotting-genomic-ranges">Now plotting genomic ranges</h3>
<pre><code>region_mod_summary &lt;- m6A_subset %&gt;%
  group_by(region, mod) %&gt;%
  summarise(n_mods = n(), .groups = "drop")

filtered_region_mod_summary &lt;- region_mod_summary %&gt;%
  filter(mod %in% c("pseU", "m6A"))

#Now Actually Plot
## This can be easily scaled up with other modification types or species

ggplot(filtered_region_mod_summary, aes(x = region, y = n_mods, fill = region)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(
    title = "Modification Counts per Region (pseU &amp; m6A)",
    x = "Genomic Region",
    y = NULL,
    fill = "Genomic Region"
  ) +
  scale_fill_aaas() +
  facet_wrap(~ mod, scales = "free_y") +
  theme_minimal(24) +
  theme(
    plot.title = element_text(face = "bold", hjust = 0.5),
    axis.text.x = element_blank()  # this removes tick labels under bars
  )</code></pre>
<p><img width="800" height="600" alt="githubexample_genomicregions" src="https://github.com/user-attachments/assets/fd1854b8-5856-433d-ab47-8c1002879210"></p>
</section>
<section id="go-enrichment-analyses" class="level3">
<h3 class="anchored" data-anchor-id="go-enrichment-analyses">GO Enrichment Analyses</h3>
<pre><code>#First load in GO term file to produce the background

background &lt;- read_delim("arabidopsis_go_terms.txt", 
delim = "\t", escape_double = FALSE,  
trim_ws = TRUE) %&gt;% 
dplyr::select(3,1) %&gt;% 
dplyr::rename(term = 1, gene = 2)

m6a_genes &lt;- final_filtered_data %&gt;%
   filter(mod == "m6A") %&gt;%
   distinct(gene_id)
m6a_genes &lt;- m6a_genes %&gt;%
  mutate(gene_id = sub("^gene:", "", gene_id))
ego &lt;- enricher(
  m6a_genes$gene_id,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  universe = background$gene,
  minGSSize = 10,
  maxGSSize = 500,
  qvalueCutoff = 0.05,
  gson = NULL,
  TERM2GENE =  goterms
  )


hyper_ego &lt;- mutate(ego, FoldEnrichment = parse_ratio(GeneRatio) / parse_ratio(BgRatio))

dotplot(hyper_ego, x="FoldEnrichment") +
  ggtitle("m6A modified genes") +
  theme_cowplot(16)</code></pre>
<p><img width="800" height="600" alt="GO_no_methylation" src="https://github.com/user-attachments/assets/b8ff7677-7e77-488a-81bf-4e598e9aa57e"></p>
</section>
<section id="next-export-the-subset-file-as-bed-to-run-bedtools-intersect-must-be-done-unix-to-line-up-with-reference-file-and-output-as-fasta" class="level3">
<h3 class="anchored" data-anchor-id="next-export-the-subset-file-as-bed-to-run-bedtools-intersect-must-be-done-unix-to-line-up-with-reference-file-and-output-as-fasta">Next export the subset file as bed to run bedtools intersect (must be done unix) to line up with reference file and output as fasta</h3>
<pre><code>m6A_subset %&gt;%
  filter(code == "a") %&gt;%
  dplyr::select(chrom, start, end, code, n_valid_cov, strand) %&gt;%
  distinct(chrom, start, end, .keep_all = T) %&gt;%
  arrange(chrom, start) %&gt;%
  mutate(start = start - 3,
         end = end + 3) %&gt;%
  write.table("C:/extract/to/bedfile.bed",
              col.names = F, row.names = F, sep = "\t", quote = F)</code></pre>
</section>
<section id="next-code-can-not-be-run-in-rstudio-must-export-to-a-unix-shell" class="level3">
<h3 class="anchored" data-anchor-id="next-code-can-not-be-run-in-rstudio-must-export-to-a-unix-shell">Next code can NOT be run in RStudio, must export to a UNIX shell</h3>
<p>The next part must be run on Unix with bedtools installed to run the intersect on the bed file with the known fasta reference</p>
<pre><code>./bedtools getfasta -s -fi "reference_toplevel.fa" -bed "C:/extract/to/bedfile.bed" &gt; "bedfile_nowfa.fasta"

samtools faidx bedfile_nowfa.fasta"</code></pre>
</section>
<section id="using-pa-lengths-from-original-basecalling-for-analyses" class="level3">
<h3 class="anchored" data-anchor-id="using-pa-lengths-from-original-basecalling-for-analyses">Using pA lengths from original basecalling for analyses</h3>
<pre><code>m6A_subset_tails &lt;- m6A_subset %&gt;%
  mutate(
    tail_length_group = case_when(
      mean_tail_length &lt;= quantile(mean_tail_length, 0.25, na.rm = TRUE) ~ "short",
      mean_tail_length &gt;= quantile(mean_tail_length, 0.75, na.rm = TRUE) ~ "long",
      TRUE ~ "medium"
    )
  )</code></pre>
<p>After all library prep is done can plot multiple variables against, like plotting CPM, stoich, and polyA length (This graph has no correlation due to the subset but minor trends can be viewed with biological data)</p>
<pre><code>ggplot(m6A_subset, aes(x = mean_CPM, y = tail_length)) +
  geom_point(alpha = 0.6) +
  geom_smooth(method = "lm", se = TRUE) +
  scale_x_log10() +
  labs(title = "Expression vs PolyA Tail Length (All Species)", x = "Mean CPM (log scale)", y = "Mean PolyA Tail Length") +
  theme_minimal(base_size = 14)</code></pre>
<p><img width="800" height="600" alt="githubexample_CPMSTOICH_AND_POLYA" src="https://github.com/user-attachments/assets/db10c5fd-0015-4d89-a92a-318c0db87ab9"></p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/gjandebeur\.github\.io\/Portfolio\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>